src/lib/ams.c

```diff
@@ -0,0 +1,749 @@
+		/* Search for first invalid entry */
+		if ((producer_routing_table[iter].message_type_id == message_type_id) &&
+		    (producer_routing_table[iter].producer_instance_id == instance_id) &&
+		    (producer_routing_table[iter].producer_module_id == module_id)) {
+			producer_routing_table[iter].message_type_id = AM_INVALID_MSG_TYPE;
+			break;
+		}
+	}
+	ams_release(ams->am_service_context);
+	return 0;
+}
+
+int am_service_register_consumer(uint32_t message_type_id,
+				 uint16_t module_id,
+				 uint16_t instance_id,
+				 ams_msg_callback_fn function,
+				 void *ctx)
+{
+	struct async_message_service *ams = sof_get()->ams;
+	struct ams_routing_entry *routing_table;
+	int err = -EINVAL;
+
+	if (ams->am_service_context == NULL || function == NULL)
+		return -EINVAL;
+
+	ams_acquire(ams->am_service_context);
+
+	routing_table = ams->am_service_context->shared->routing_table;
+	for (uint32_t iter = 0; iter < ams->am_service_context->shared->routing_table_size; iter++) {
+		/* Search for first invalid entry */
+		if (routing_table[iter].message_type_id == AM_INVALID_MSG_TYPE) {
+			/* Add entry to routing table for local service */
+			routing_table[iter].consumer_callback = function;
+			routing_table[iter].message_type_id = message_type_id;
+			routing_table[iter].consumer_instance_id = instance_id;
+			routing_table[iter].consumer_module_id = module_id;
+			routing_table[iter].consumer_core_id = cpu_get_id();
+			routing_table[iter].ctx = ctx;
+
+			/* Exit loop since we added new entry */
+			err = 0;
+			break;
+		}
+	}
+
+	ams_release(ams->am_service_context);
+	return err;
+}
+
+int am_service_unregister_consumer(uint32_t message_type_id,
+				   uint16_t module_id,
+				   uint16_t instance_id,
+				   ams_msg_callback_fn function)
+{
+	struct async_message_service *ams = sof_get()->ams;
+	struct ams_routing_entry *routing_table;
+	int err = -EINVAL;
+
+	if (ams->am_service_context == NULL)
+		return -EINVAL;
+
+	ams_acquire(ams->am_service_context);
+	routing_table = ams->am_service_context->shared->routing_table;
+	for (uint32_t iter = 0; iter < ams->am_service_context->shared->routing_table_size; iter++) {
+		/* Search for required entry */
+		if ((routing_table[iter].message_type_id == message_type_id) &&
+		    (routing_table[iter].consumer_module_id == module_id) &&
+		    (routing_table[iter].consumer_instance_id == instance_id) &&
+		    (routing_table[iter].consumer_callback == function)) {
+			/* Remove this entry from routing table */
+			routing_table[iter].message_type_id = AM_INVALID_MSG_TYPE;
+			routing_table[iter].consumer_callback = NULL;
+
+			/* Exit loop since we removed entry */
+			err = 0;
+			break;
+		}
+	}
+
+	ams_release(ams->am_service_context);
+	return err;
+}
+
+uint32_t am_service_push_slot(struct am_service_context *const am_service_ctx,
+			      const struct ams_message_payload *msg,
+			      uint16_t module_id, uint16_t instance_id)
```

Comment:
shouldn't be unsigned - it can return negative errors. Please, consider using standard C types like `int` where the exact size isn't important